src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  const stateCopy = { ...state };
+
+  for (let i = 0; i < actions.length; i++) {
+    if (actions[i].type === 'addProperties') {
+      Object.assign(stateCopy, actions[i].extraData);
+      history.push({ ...stateCopy });
+    }
+
+    if (actions[i].type === 'removeProperties') {
+      for (let j = 0; j < actions[i].keysToRemove.length; j++) {
+        delete stateCopy[actions[i].keysToRemove[j]];
+      }
+      history.push({ ...stateCopy });
+    }
+
+    if (actions[i].type === 'clear') {
+      Object.keys(stateCopy).forEach((key) => delete stateCopy[key]);
+      history.push({ ...stateCopy });
```

Comment:
[DONT REPEAT YOURSELF]: You're pushing a copy of the state to the history array in every case. To avoid repetition, perform this action once after the switch statement at the end of each iteration.